File: 2018/05-polymer-reactions/go/star1.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"
)

func main() {
	input, _ := ioutil.ReadFile("input")
	polymer := strings.TrimSpace(string(input))

	var stack []rune
	for _, unit := range polymer {
		var previous rune
		if len(stack) > 0 {
			previous = stack[len(stack)-1]
		}

		// If same letter, different case
		charDiff := math.Abs(float64(unit - previous))
		if charDiff == 32.0 {
			// Pop previous
			stack = stack[:len(stack)-1]
			continue
		}

		stack = append(stack, unit)
	}

	fmt.Println(len(stack))
}

File: 2018/05-polymer-reactions/rust/star1.rs
const INPUT: &str = include_str!("../input");

fn reacts(c1: char, c2: char) -> bool {
    return (c1 != c2) && (c1.to_uppercase().to_string() == c2.to_uppercase().to_string());
}

fn main() {
    let polymer = INPUT.trim();

    let mut stack = Vec::new();
    for c in polymer.chars() {
        if stack.is_empty() {
            stack.push(c);
            continue;
        }

        if reacts(c, stack[stack.len() - 1]) {
            stack.pop();
        } else {
            stack.push(c);
        }
    }

    println!("{}", stack.len());
}
